Accept placeholder_data via URL query param

timhallmann opened the pull request:
It's quite useful for us to generate links that set a few properties and let the user enter the remaining data.

While I like the ergonomics from a user perspective, the implementation feels a bit more complicated than it should be. Am I missing any functions that could aid in the parsing of the arguments?

If there are any reasons to change the `property` key or if keys like `properties.name=test` would be preferable, I'd be happy to change that.

E.g.:

```
/objects/new?action_id=1
&property=name=test
&property=count=100
&property=size=10nm
&property=created=2025-01-01T10:20:30
&property=boolfield=false
&property=sample=1
&property=tags=a,b
&property=multilayer.0.repetitions=1
```

Note: Some of these properties are not valid for action 1 in the demo, so these would need to be removed to test this feature.

FlorianRhiem commented:
Hey @timhallmann, thank you for the pull request. Can you elaborate a bit on the use cases for this feature? Is there any overlap with the Recipes feature, or the "Use as template" feature combined with the `may_copy` property in schemas?

Would these URLs/query parameters be hand-written by regular users, by administrators, or generated in some way? The syntax you propose is fairly simple to write/read, which is a big advantage if regular users are supposed to write these. A slightly more elaborate syntax might be beneficial to writing a much simpler implementation, ideally one that is mostly independent of the property type, though.

FlorianRhiem commented:
Something like  the following code could handle parameters like this:

```
http://localhost:5000/objects/new?action_id=5
&property=name.text=%22Test%22
&property=mass.magnitude=1
&property=mass.units=%22kg%22
```

The URL-escaped double codes are a bit ugly, I used them so I could simply apply `json.loads` to the values, but the key is that if strings and other values (i.e. numbers and bools) can be told apart without looking at the schema, etc, then the code is simplified a good bit. 

```python
    try:
        url_properties = {}
        for property in flask.request.args.getlist('property'):
            url_path, _, url_value = property.partition('=')

            path = []
            for k in url_path.split('.'):
                try:
                    k = int(k)
                except:
                    pass
                path.append(k)

            s = action.schema
            for k in path[:-1]:
                while s.get('type', None) in ('object', 'array'):
                    if s['type'] == 'object':
                        s = s['properties']
                    else:
                        s = s['items']
                if isinstance(k, int):
                    continue
                s = s[k]

            d = {
                '_type': s['type'],
                path[-1]: json.loads(url_value)
            }
            if tuple(path[:-1]) not in url_properties:
                url_properties[tuple(path[:-1])] = [{}, s]
            url_properties[tuple(path[:-1])][0].update(d)
        if not placeholder_data:
            placeholder_data = {}
        for path in url_properties:
            data, schema = url_properties[path]
            logic.schemas.validate(data, schema, list(path))
            placeholder_data[path] = data
    except Exception as e:
        # Ignore invalid data
        pass
```

timhallmann commented:
My primary use case right now is to link an action from our order system and fill in the order id and one or two parameters that usually follow from the order request. No overlap with recipes or templates, but these URLs would be generated automatically.

I considered parsing `?property.path:key=value` as `path: {key: value}` or passing a Base64-encoded (partial) JSON representation of an object (or placeholder_data), but even programmatically, it seemed quite a bit more cumbersome to use, while not saving that much effort on the implementation. It would, of course, remove the need to handle every datatype differently, but I'd assume changes to this code would be rather seldom.

I decided on the current implementation when I realized I needed to set
`?property.path:magnitude_in_base_units=value` for a simple number, while a quantity with a unit would look like this: `?property.path:magnitude=value&property.path:unit=nm`.

Offering a simple user interface seemed worth it, even if it's not likely that our staff or users would end up creating URLs using this feature.

FlorianRhiem commented:
`?property.path:magnitude=value` should suffice for any quantity property with a single unit, but your point stands. In that case, perhaps the parsing can be simplified with a dict that maps the property type to the property dict key and value type that needs to be parsed, e.g. `value` and `bool` for a boolean, `text` and `str` for a text, etc. Quantities with fixed units would also work that way, only quantities with variable units would need more specific handling.

timhallmann commented:
I've simplified this using a dict and split it into a few utility functions, though I've kept them within `view.py` for now. Not sure if there's any value in splitting it into utility functions. Type hints are only partially fixed.

I've also switched to `?properties.x=1`, instead of `?property=x=1`. `properties` matches the schemas a bit better than `property`.

FlorianRhiem commented on sampledb/frontend/objects/view.py:
```diff
@@ -1254,6 +1273,61 @@ def is_property_schema_compatible(
+        d = {
+            'magnitude': magnitude,
+        }
+        if unit in ('1', ''):
+            d = {
+                'magnitude_in_base_units': value,
+            }
+        if unit:
+            d['units'] = unit
```
This seems unnecessarily complicated to me, why not use `magnitude` independent of the unit, and set the unit even if it is `1`? If, for example, the units allowed for a quantity are percent and 1, then this code might lead to ambiguous / invalid data, wouldn't it? And if `unit` is in the list of units allowed by the schema, then why not set it? Validation will take care of the rest, such as `dimensionality` and `magnitude_in_base_units`. Perhaps I'm overlooking something, though.

timhallmann replied:
`{'magnitude': '123', 'units': '1'}` does not set a default. Nor does `{'magnitude': 123.0, 'units': '1'}`. If that should happen, then it seems like there's a bug somewhere. I haven't yet looked for the relevant functions.

I hadn't thought about percents so far. But always setting `magnitude` and `units` should cover percents as well, assuming that works for `1`, right?

FlorianRhiem replied:
Are you validating the data you generate there? That's the step that adds the additional, missing values.

The issue with `percent` that I meant is that if the quantity schema allows both `1` and `percent` as units, then a unit should be specified in the data, and since `1` as a unit is fine, I don't see why not to set it.

timhallmann replied:
I didn't add validation steps. Should that be added to `show_object_form` after `_apply_placeholder_data` using `logic.schemas.validate` or do you mean the `placeholder_data` itself or another function?

Yeah, I intend to remove the special case entirely, if that should work.

FlorianRhiem replied:
You should be able to add it to the end of `parse_default_for_subschema`, by storing the data in a variable instead of returning it right away, and then passing that variable to `logic.schemas.validate(data, schema, path)`. Alternatively, you could iterate over `url_properties` and validate, like I did in my snippet, though you'd also need to store the schemas for each path as well, then.

timhallmann replied:
Oh, sorry, I didn't look that closely into your snippet.

I've updated the commit. Schemas may not use `"type": ""` to declare a dimensionless quantity, they must use `1`.

FlorianRhiem commented:
I've done some manual tests and added some [tests](https://github.com/FlorianRhiem/sampledb/commit/c57c3e9c5579540c30bf3a0765dcf901d775459d) in my fork, the implementation seems to work nicely. Could you document the feature?

timhallmann commented:
Thank you very much for the tests! Can I cherry-pick your commit to my branch or would you prefer to keep that separate?

Yeah, I can document that. In "Metadata and Schemas" or somewhere else?

I've noticed another issue: How should we handle values which have a different unit than the unit(s) in the schema, but do have the same dimensionality? I assumed we'd only get a problem with `""` and `"1"`, because these were the only ones with special cases, but we get same issue with `"%"` and `"percent"`, `"µm"` and `"um"`, etc.

We could validate only the dimensionality via `logic.schemas.validate`, but then we run into the following: If a field allows only a single unit and we pass a different unit of the same dimensionality, the user isn't offered the option to change the unit (e.g. schema says `"s"` and we pass `?properties.time=5day`). `quantity_field.html` would need to also compare the `selected_unit` with the unit(s) in the schema. Preferably while also handling aliases, otherwise we'd list the same unit twice.

FlorianRhiem commented:
Sure, you can cherry-pick it.

The current implementation of only accepting units that are permitted by the schema seems fine, to me. `quantity_field.html` does actually perform that comparison, however it is string based. Yes, this would lead to duplicate units if we accept any unit, e.g. if `deg` was listed in the schema but `degree` was given, but this doesn't occur if we keep the implementation so that it uses the units given by the schema. If you are worried about this being unintuitive for manually building these URLs, then I'd suggest flashing error messages for invalid query parameters, that way the message for an invalid unit could contain the valid units?

FlorianRhiem commented:
Looks good to me. The CI pipeline complains about the formatting of the translations, [here's a commit that wraps the lines correctly](https://github.com/FlorianRhiem/sampledb/commit/676939ab15b6cd0f79e356b7168ac8fe87f6ec6f). Perhaps squash that into your commit, alternatively I can also merge this with that extra commit in place.

timhallmann commented:
Done.

Thanks again for your work on this patch!

FlorianRhiem commented:
Thank you for suggesting and implementing this feature! I've merged it into develop. Perhaps you could write a post about it to the mailing list? You can probably stick quite close to the documentation, but chances are better that people will see it that way than by coming across it in the docs/changelog :)
